### PR TITLE
C#: Update to support .NET Core 2.1

### PR DESCRIPTION
**Pull request by natemcmaster**
Support the new package naming scheme for Debian packages.

Resolves https://github.com/travis-ci/travis-ci/issues/9684

cc @Joshua-Anderson @akoeplinger @nterry

### CODE REVIEW

**Comment by natemcmaster** [comment]
Yeah, full version for macOS is the best way forward for now. We don't have a semantic way to express '2.1' in macOS downloads yet.

**Comment by akoeplinger** [comment]
@BanzaiMan let's go with this PR instead :)

File: lib/travis/build/script/csharp.rb [hunk 1]
```diff
@@ -321,10 +330,26 @@ def is_mono_after_5_0
...
           return false unless is_dotnet_version_valid?
 
-          return false if DOTNET_VERSION_REGEXP.match(config_dotnet)[1].to_i < 2
+          return false if DOTNET_VERSION_REGEXP.match(config_dotnet)[:major].to_i < 2
 
           true
         end
+
+        # Starting in 2.1.300, the package ID changed from dotnet-sdk-(major.minor.patch) to dotnet-sdk-(major.minor)
+        def is_dotnet_after_2_1_300?
```
Comment (BanzaiMan) [inline lib/travis/build/script/csharp.rb]: You can use `Gem::Version` to compare version numbers. See, for example, https://github.com/travis-ci/travis-build/blob/a79566b678464c80cb613f8636cbe89114d623a0/lib/travis/build/script/elixir.rb#L88-L91

Comment (natemcmaster) [reply lib/travis/build/script/csharp.rb]: Awesome, that's much cleaner. Just updated to use that.

**Comment by BanzaiMan** [comment]
This has been deployed to https://staging.travis-ci.org. Please test it.

**Comment by natemcmaster** [comment]
How can I setup a repo in the staging environment? I couldn't login or find instructions on testing with this environment.

**Comment by BanzaiMan** [comment]
@natemcmaster 
> I couldn't login or find instructions on testing with this environment.

What do you see in https://staging.travis-ci.org?

**Comment by natemcmaster** [comment]
After clicking 'sign in with github' on https://staging.travis-ci.org, it redirects to https://docs.travis-ci.com/api.

**Comment by BanzaiMan** [comment]
I see. The UI is being developed there, so we'll have to try again later.

**Comment by natemcmaster** [comment]
Ok. I can't help test this then. But here's a repo you could fork and
build. If it passes, then we're good.
https://github.com/natemcmaster/CommandLineUtils/blob/master/.travis.yml

**Comment by backspace** [comment]
Hey, @natemcmaster, we found that we had a misconfiguration but staging.travis-ci.org is back to working as expected now.

**Comment by fubar-coder** [comment]
@backspace I tried to use it, but it couldn't find the branches for the fluentmigrator/fluentmigrator repository.

**Comment by natemcmaster** [comment]
https://staging.travis-ci.org/natemcmaster/CommandLineUtils/builds/722430

Apt-get issued the right command, but it appears to have failed due do
network issues.

**Comment by BanzaiMan** [comment]
@fubar-coder Could you elaborate? Did you try pushing a commit to it? Or did you just try to trigger a build after enabling it?

**Comment by fubar-coder** [comment]
@BanzaiMan Everything works now after I manually triggered a build. This is the error I get now:

```text
/usr/share/dotnet/sdk/2.1.300/NuGet.targets(114,5): error : No space left on device [/home/travis/build/fluentmigrator/fluentmigrator/FluentMigrator.sln]
```

**Comment by BanzaiMan** [comment]
Please test something below 2.1.300.

**Comment by BanzaiMan** [comment]
@fubar-coder Try syncing data.

**Comment by BanzaiMan** [comment]
If that doesn't help, please push a commit.

**Comment by fubar-coder** [comment]
Installing an older version of .NET Core did work (2.1.103). Syncing, Disabling and Re-enabling the project didn't work, so I created a `test` branch which was picked up and triggered a build.

**Comment by BillChirico** [comment]
When can we expect this fix to be in production?

**Comment by BanzaiMan** [comment]
I've just deployed it.

**Comment by JustArchi** [comment]
Side note: `2.1` doesn't work for OS X, `2.1.300` works on both. It might make sense to add some compatibility for that if you're interested.

**Comment by natemcmaster** [comment]
Yes, this is a known limitation. I'm not quite sure how to resolve it though. There is no apt-get equivalent in macOS which can tell us easily which is the latest `2.1` thing to fetch.

**Comment by JustArchi** [comment]
Can't you (as a .NET Core team) add some symlink or other compatibility layer that would point to proper release? I mean, this is inconsistent behaviour which in fact .NET Core devs should be interested in, assuming they consider it as such that is.

**Comment by natemcmaster** [comment]
@JustArchi let's use https://github.com/dotnet/core/issues/1649 to follow up on this.

**Pull request merged**